src/databricks/sql/thrift_backend.py

```diff
@@ -279,24 +283,40 @@ def attempt_request(attempt):
             # - non-None method_return -> success, return and be done
             # - non-None retry_delay -> sleep delay before retry
             # - error, error_message always set when available
+
+            error = None
             try:
                 logger.debug("Sending request: {}".format(request))
                 response = method(request)
                 logger.debug("Received response: {}".format(response))
                 return response
-            except Exception as error:
+            except socket.timeout as err:
+                # socket.timeout means the connection was timed out by the socket package
+                retry_delay = self._retry_delay_default
+                error_message = str(err)
+                error = err
+            except OSError as err:
+                # OSError 110 means the connection was timed out by the operating system
+                if "Errno 110" in str(err):
+                    retry_delay = self._retry_delay_default
+                else:
+                    retry_delay = None
+                error_message = str(err)
+                error = err
+            except Exception as err:
                 retry_delay = extract_retry_delay(attempt)
```

Comment:
What does this mean? It means retry on exceptions in general?

Reply:
No, if `retry_delay` is `None` then the request is not retried. `extract_retry_delay` returns `None` in all cases except for a 429/503 code with `Retry-After` headers sent.

In general, requests are not considered retryable.

Meta point: this logic needs a full reworking for clarity as part of a follow-up ticket. The retry logic is only going to become more complicated from here as we unify the python connector behaviour with the ODBC behaviour

Reply:
The reason for these extra `except` statements is for the case when a request should be retried but won't have `Retry-After` headers because a response was never received from the server.